File: src/ai_guardrails/writers/sarif.py
"""Build SARIF 2.1.0 output from LintIssue findings."""

from __future__ import annotations

from importlib.metadata import version as _pkg_version
import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ai_guardrails.models.lint_issue import LintIssue

_SCHEMA = "https://json.schemastore.org/sarif-2.1.0.json"
_TOOL_VERSION = _pkg_version("ai-guardrails")


def build_sarif(issues: list[LintIssue], *, tool_version: str = _TOOL_VERSION) -> str:
    """Return SARIF 2.1.0 JSON string for the given lint issues."""
    rules = {i.rule: {"id": i.rule, "name": i.rule} for i in issues}
    results = [
        {
            "ruleId": i.rule,
            "level": "error",
            "message": {"text": i.message},
            "locations": [
                {
                    "physicalLocation": {
                        "artifactLocation": {"uri": i.file},
                        "region": {"startLine": i.line, "startColumn": i.col},
                    }
                }
            ],
        }
        for i in issues
    ]
    sarif = {
        "$schema": _SCHEMA,
        "version": "2.1.0",
        "runs": [
            {
                "tool": {
                    "driver": {
                        "name": "ai-guardrails",
                        "version": tool_version,
                        "rules": list(rules.values()),
                    }
                },
                "results": results,
            }
        ],
    }
    return json.dumps(sarif, indent=2)

File: src/writers/sarif.ts
import type { LintIssue } from "@/models/lint-issue";

interface SarifLocation {
  physicalLocation: {
    artifactLocation: { uri: string };
    region: { startLine: number; startColumn: number };
  };
}

interface SarifResult {
  ruleId: string;
  level: "error" | "warning" | "note";
  message: { text: string };
  locations: SarifLocation[];
}

interface SarifRun {
  tool: { driver: { name: string; version: string; rules: unknown[] } };
  results: SarifResult[];
}

interface SarifLog {
  version: "2.1.0";
  $schema: string;
  runs: SarifRun[];
}

function severityToLevel(severity: LintIssue["severity"]): SarifResult["level"] {
  return severity === "error" ? "error" : "warning";
}

/**
 * Convert LintIssue[] to SARIF 2.1.0 format.
 */
export function issuesToSarif(issues: LintIssue[]): SarifLog {
  const results: SarifResult[] = issues.map((issue) => ({
    ruleId: issue.rule,
    level: severityToLevel(issue.severity),
    message: { text: issue.message },
    locations: [
      {
        physicalLocation: {
          artifactLocation: { uri: issue.file },
          region: { startLine: issue.line, startColumn: issue.col },
        },
      },
    ],
  }));

  const run: SarifRun = {
    tool: {
      driver: {
        name: "ai-guardrails",
        version: "3.0.0",
        rules: [],
      },
    },
    results,
  };

  return {
    version: "2.1.0",
    $schema:
      "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
    runs: [run],
  };
}
